### PR TITLE
Report tests to Datadog from Job that executed them

### PR DESCRIPTION
**Pull request by eps1lon**
TL;DR
```diff
-@test_session.name:report-test-results-to-datadog-datadog-ci junit upload
+@test_session.name:test-prod-react--4/7
```

Previously, each test run was associated with a separate Job that is responsible for uploading all reports.
This makes it harder to understand under which configuration a specific test failed.
Now we immediately upload test results from the respective job meaning a test run in Datadog can be associated with the actual CI job that ran it and therefore it's config (Turbopack vs Webpack, PPR vs no PPR, React 18 vs 19)

This increases feedback time slightly by up to 1 minute (the slowest single upload time).
I'd say this is an acceptable tradeoff given our current CI times.

This reduces the time until a "Retry failed jobs" is avaiable since we no longer need to pass the test results between jobs.
 
Tried setting `DD_GITHUB_JOB_NAME` and `DD_TEST_SESSION_NAME` without success: https://github.com/DataDog/datadog-ci/issues/1497

Before:
CleanShot 2025-02-18 at 21 21 31

-- https://app.datadoghq.com/ci/test/runs?query=test_level%3Atest%20%40git.branch%3A%22mischnic%2Fasync-info-module-graph%22&agg_m=count&agg_m_source=base&agg_t=count&cols=%40test.status%2C%40test_session.name%2C%40test.name%2C%40test.source.file%2C%40test.service%2C%40test.type%2C%40ci.pipeline.name%2C%40duration&currentTab=overview&eventStack=&fromUser=true&graphType=flamegraph&index=citest&start=1739305301799&end=1739910101799&paused=true

After:
CleanShot 2025-02-18 at 23 13 01

-- https://app.datadoghq.com/ci/test/runs?query=test_level%3Atest%20%40ci.job.url%3A%22https%3A%2F%2Fgithub.com%2Fvercel%2Fnext.js%2Fcommit%2Fe785d486b106c745a89caa72188ef0446f87c1f1%2Fchecks%22&agg_m=count&agg_m_source=base&agg_t=count&cols=%40test_session.name%2C%40test.status%2C%40test.name%2C%40test.source.file%2C%40test.service%2C%40test.type%2C%40ci.pipeline.name%2C%40duration&currentTab=overview&eventStack=&fromUser=true&graphType=flamegraph&index=citest&start=1739311985833&end=1739916785833&paused=true

### CODE REVIEW

**Comment by eps1lon** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/vercel/next.js/73180?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#73180** <a href="https://app.graphite.dev/github/pr/vercel/next.js/73180?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/vercel/next.js/73180?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#73006** <a href="https://app.graphite.dev/github/pr/vercel/next.js/73006?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#72996** <a href="https://app.graphite.dev/github/pr/vercel/next.js/72996?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#73000** <a href="https://app.graphite.dev/github/pr/vercel/next.js/73000?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#72999** <a href="https://app.graphite.dev/github/pr/vercel/next.js/72999?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: .github/workflows/build_reusable.yml [hunk 1]
```diff
@@ -81,6 +81,11 @@ env:
   DATADOG_API_KEY: ${{ secrets.DATA_DOG_API_KEY }}
   NEXT_JUNIT_TEST_REPORT: 'true'
   DD_ENV: 'ci'
+  DD_CI_JOB_NAME: ${{ inputs.stepName }}
+  # Better than Datadog's default (e.g. https://github.com/vercel/next.js/commit/3dd3d19ba63c7ca790f3bf39e0e15152b597547c/checks)
+  # Job id is not provided by GitHub. Using workflow url as a fallback.
+  # We could derived the job id from GH API response but this is just more network slowness.
+  DD_CI_JOB_URL: https://github.com/${{  github.repository }}/actions/runs/${{ github.run_id }}/attempts/${{ github.run_attempt }}
```
Comment (eps1lon) [inline .github/workflows/build_reusable.yml:88]: Ignored for unknown reasons: https://github.com/DataDog/datadog-ci/issues/1497

**Review by ijjk — APPROVED** [review]
I'm seeing average upload times around `10s` so this seems alright to try if it gives us better insights
